File: app/src/main/java/com/cameraviewer/app/TailscaleDiscovery.kt
package com.cameraviewer.app

import android.util.Base64
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

data class TailscalePeer(val hostname: String, val ipv4: String)

/**
 * Lists tailnet devices via the Tailscale REST API, same approach as the
 * laptop-side discovery design for the web dashboard — except this runs
 * directly against api.tailscale.com over the internet rather than the
 * `tailscale` CLI, since Termux/Android has no CLI to shell out to.
 *
 * Auth: API access tokens (tskey-api-...) are presented as HTTP Basic Auth
 * with the token as the username and an empty password — NOT a bearer
 * token. This is easy to get wrong; Tailscale's docs use curl's `-u
 * "$TOKEN:"` form, which is exactly this.
 */
object TailscaleDiscovery {
    private const val TAG = "TailscaleDiscovery"
    private const val DEVICES_URL = "https://api.tailscale.com/api/v2/tailnet/-/devices"

    suspend fun listPeers(apiToken: String): List<TailscalePeer> = withContext(Dispatchers.IO) {
        val conn = URL(DEVICES_URL).openConnection() as HttpURLConnection
        try {
            val basic = Base64.encodeToString("$apiToken:".toByteArray(), Base64.NO_WRAP)
            conn.setRequestProperty("Authorization", "Basic $basic")
            conn.connectTimeout = 10_000
            conn.readTimeout = 10_000

            val code = conn.responseCode
            if (code != HttpURLConnection.HTTP_OK) {
                throw TailscaleApiException(
                    "Tailscale API returned HTTP $code" +
                        if (code == 401 || code == 403) " — check the API token in Settings" else "",
                )
            }

            val body = conn.inputStream.bufferedReader().use { it.readText() }
            val devices = JSONObject(body).getJSONArray("devices")
            val peers = mutableListOf<TailscalePeer>()
            for (i in 0 until devices.length()) {
                val device = devices.getJSONObject(i)
                val hostname = device.optString("hostname", device.optString("name", "unknown"))
                val addresses = device.optJSONArray("addresses") ?: continue
                var ipv4: String? = null
                for (j in 0 until addresses.length()) {
                    val addr = addresses.getString(j)
                    // Tailscale IPv4 addresses are always in the 100.64.0.0/10 CGNAT range.
                    if (addr.startsWith("100.")) {
                        ipv4 = addr
                        break
                    }
                }
                if (ipv4 != null) {
                    peers.add(TailscalePeer(hostname, ipv4))
                }
            }
            Log.i(TAG, "discovered ${peers.size} tailnet peer(s) with an IPv4 address")
            peers
        } finally {
            conn.disconnect()
        }
    }
}

class TailscaleApiException(message: String) : Exception(message)

File: desktop-viewer/tailscale_discovery.py
"""Lists tailnet devices via the Tailscale REST API — same endpoint and
auth scheme as the Android app's TailscaleDiscovery.kt, so a token that
works in the phone app's Settings works here unchanged.

Auth: API access tokens (tskey-api-...) are HTTP Basic Auth with the token
as the username and an empty password (requests' `auth=(token, "")`) — NOT
a bearer token. Easy to get wrong; Tailscale's own docs use curl's
`-u "$TOKEN:"` form, which is exactly this.
"""

from __future__ import annotations

import socket

import requests

DEVICES_URL = "https://api.tailscale.com/api/v2/tailnet/-/devices"
REQUEST_TIMEOUT = 10
PROBE_TIMEOUT = 2

RELAY_PORT = 8792  # VideoRelayServerService
ALERT_PORT = 8790  # AlertReceiverService


class TailscaleApiError(Exception):
    pass


def list_peers(api_token: str) -> list[dict]:
    """Returns [{"hostname": str, "ip": str}, ...] for every tailnet device with an IPv4 address."""
    try:
        resp = requests.get(DEVICES_URL, auth=(api_token, ""), timeout=REQUEST_TIMEOUT)
    except requests.RequestException as e:
        raise TailscaleApiError(f"could not reach Tailscale API: {e}") from e

    if resp.status_code != 200:
        hint = " — check the API token in Settings" if resp.status_code in (401, 403) else ""
        raise TailscaleApiError(f"Tailscale API returned HTTP {resp.status_code}{hint}")

    peers = []
    for device in resp.json().get("devices", []):
        hostname = device.get("hostname") or device.get("name") or "unknown"
        ipv4 = next((a for a in device.get("addresses", []) if a.startswith("100.")), None)
        if ipv4:
            peers.append({"hostname": hostname, "ip": ipv4})
    return peers


def probe_port(ip: str, port: int, timeout: float = PROBE_TIMEOUT) -> bool:
    """Weak discovery signal (matches ViewerProber's approach): just checks something is listening. Same caveat applies — this can false-positive on any other device with the port open, unlikely on a home tailnet."""
    try:
        with socket.create_connection((ip, port), timeout=timeout):
            return True
    except OSError:
        return False


def scan_for_cameras(api_token: str) -> list[dict]:
    """Tailnet peers with an open video-relay port — candidate sender phones."""
    peers = list_peers(api_token)
    return [p for p in peers if probe_port(p["ip"], RELAY_PORT)]
